File: 백준/Silver/9012. 괄호/괄호.cc
#include<bits/stdc++.h>
using namespace std;
int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int n;
    cin >> n;
    while (n--) {
        stack<char> s;
        string g;
        cin >> g;
        bool at = true;
        for (auto c : g) {
            if (c == '(') s.push(c);
            else {
                if (s.empty() || s.top() == ')') {
                    at = false;
                    break;
                }
                else s.pop();
            }
        }
        if (at && s.empty()) cout << "YES\n";
        else cout << "NO\n";
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
import sys
input = sys.stdin.readline

t = int(input())
for _ in range(t):
    stack = []
    a = input().strip()
    for i in a:
        if i == '(':
            stack.append(i)
        elif i ==')' and stack:
            stack.pop()
        else:
            stack.append(i)
            break
    if stack:
        print('NO')
    else:
        print('YES')
                